File: backend/models/import.go
package models

import (
	"time"

	"mycorrhizal/contactmodel"
)

// ImportableContactFields defines the valid target fields for import
var ImportableContactFields = []string{
	// Scalars
	"firstname", "lastname", "middle_name", "prefix", "suffix", "nickname", "gender",
	"birthday", "anniversary", "organization", "department", "job_title", "role",
	"how_we_met", "work_information", "contact_information",
	// Groupings. Separate targets since T3: "circles" materializes Circle +
	// CircleMember rows, "tags" materializes Tag + ContactTag rows. Before
	// that, tag-shaped headers were folded into "circles" because Tag had no
	// destination.
	"circles", "tags",
	// Multi-value values
	"email", "phone", "url", "impp",
	"address_street", "address_city", "address_region", "address_postal", "address_country",
	// Multi-value labels/types
	"email_label", "phone_label", "url_label", "impp_label", "address_label",
}

// ColumnMapping represents how a CSV column maps to a contact field
type ColumnMapping struct {
	CSVColumn    string `json:"csv_column" validate:"required"`
	ContactField string `json:"contact_field"` // Empty means "ignore this column"
	Group        int    `json:"group"`         // Multi-value entry index (0-based); ties value+label/parts of one entry together
}

// ImportUploadResponse is returned after CSV upload
type ImportUploadResponse struct {
	SessionID         string          `json:"session_id"`
	Headers           []string        `json:"headers"`
	SuggestedMappings []ColumnMapping `json:"suggested_mappings"`
	RowCount          int             `json:"row_count"`
	SampleData        [][]string      `json:"sample_data"` // First few rows for preview
}

// ImportPreviewRequest is sent to request a preview with mappings
type ImportPreviewRequest struct {
	SessionID string          `json:"session_id" validate:"required"`
	Mappings  []ColumnMapping `json:"mappings" validate:"required,dive"`
}

// DuplicateMatch describes a potential duplicate contact
type DuplicateMatch struct {
	ExistingContactID uint   `json:"existing_contact_id"`
	ExistingFirstname string `json:"existing_firstname"`
	ExistingLastname  string `json:"existing_lastname"`
	ExistingEmail     string `json:"existing_email"`
	ExistingPhone     string `json:"existing_phone"`
	MatchReason       string `json:"match_reason"` // "email", "name", or "phone"
}

// ImportScalarChange is one scalar field the "Merge" (update) action will
// overwrite on the matched existing contact, reported as a before/after pair
// (T96 — T96).
// Field is a stable camelCase key so the client can group/label by it; Label
// is the human-readable fallback the server already knows.
type ImportScalarChange struct {
	Field string `json:"field"`
	Label string `json:"label"`
	Old   string `json:"old"`
	New   string `json:"new"`
}

// ImportAddedValue is one multi-valued entry the "Merge" action will append
// to the matched existing contact — a phone/email/address/url/impp the
// existing contact does not already carry. Kind is a stable lowercase token
// ("email", "phone", "address", "url", "impp"); Value is a display rendering.
type ImportAddedValue struct {
	Kind  string `json:"kind"`
	Value string `json:"value"`
}

// ImportMergeDiff describes, per duplicate row, exactly what the "Merge"
// (update) action will change on the matched existing contact: which scalars
// get overwritten (incoming-wins-when-non-empty, the MergeImportedContact
// policy) and which multi-valued entries get appended (additive, T49). It is
// computed by services.ComputeImportMergeDiff from the same helpers the
// confirm path applies, so the preview can never describe a merge that the
// commit would not perform. Updated and Added deliberately have no omitempty:
// absent-vs-`[]` must stay distinguishable (CLAUDE.md frontend trap #8) for
// the clients that render the diff.
type ImportMergeDiff struct {
	Updated []ImportScalarChange `json:"updated"`
	Added   []ImportAddedValue   `json:"added"`
}

// ImportRowPreview represents one row in the import preview
type ImportRowPreview struct {
	RowIndex         int                    `json:"row_index"`
	ParsedContact    map[string]interface{} `json:"parsed_contact"`    // Parsed field values
	ValidationErrors []string               `json:"validation_errors"` // Any validation issues
	DuplicateMatch   *DuplicateMatch        `json:"duplicate_match"`   // Potential duplicate, if any
	SuggestedAction  string                 `json:"suggested_action"`  // "add", "skip", or "update"

	// MergeDiff is present exactly when DuplicateMatch is: what "update"
	// would change on the matched existing contact (see ImportMergeDiff).
	// Absent (nil) for rows with no duplicate or rows whose existing contact
	// could not be loaded.
	MergeDiff *ImportMergeDiff `json:"merge_diff"`

	// BatchDuplicateOf, when non-nil, holds the row_index of an EARLIER row in
	// the SAME import file that this row duplicates (T96's within-batch
	// detection — the same person imported twice in one file). These rows
	// default to "skip"; the user may still override to Keep Both.
	BatchDuplicateOf *int `json:"batch_duplicate_of"`

	// Diagnostics surfaces contactmodel.Diagnostic events  from
	// the vcard4/vcard3/jscontact adapter that parsed this row — e.g. a
	// gracefully-dropped, no-target-home field (docs/adrs/0001-neutral-hub-and-spoke-contact-model.md
	// degradation policy). Structured, carrying the same severity/concept/message
	// shape the export loss reports use (DATA-02, issue #442) — never a
	// flattened string copy. Empty for CSV-import rows (which don't go
	// through an adapter at all).
	Diagnostics []contactmodel.Diagnostic `json:"diagnostics,omitempty"`

	// CustomFieldCandidates lists the unknown X-* vCard/JSContact properties
	// discovered in this row's passthrough that the import wizard can promote
	// to first-class custom fields (issue #514) — the "promotion" step that
	// turns opaque passthrough data into editable/searchable/merge-safe
	// FieldValues. Absent (nil) for CSV/records-import rows (no adapter
	// passthrough) and for adapter rows that carried no X-* properties.
	// Matches the Diagnostics field's omitempty convention: an absent key
	// means "no candidates".
	CustomFieldCandidates []DiscoveredProperty `json:"custom_field_candidates,omitempty"`
}

// DiscoveredProperty is one unknown X-* vCard/JSContact property found on an
// imported contact that can be promoted to a custom field (issue #514).
// Surfaced on ImportRowPreview (per row) so the import wizard can offer the
// mapping step; the mapping decision itself is sent back as an
// ImportFieldMapping on the confirm request.
type DiscoveredProperty struct {
	// Name is the vCard property name as stored in Passthrough (lowercase per
	// the vcard3/vcard4 adapters' fieldToJCardProp, e.g. "x-favorite-color").
	Name string `json:"name"`
	// Value is a human-readable rendering of the property's value (the first
	// value when the property repeats).
	Value string `json:"value"`
	// MatchedDefinitionID, when non-empty, is an existing FieldDefinition
	// whose vcard:X-<NAME> projection matches this property name
	// (case-insensitively). The wizard pre-selects "map to this definition",
	// and absent a wizard override the confirm path auto-promotes
	// (the round-trip guarantee for exported custom fields).
	MatchedDefinitionID string `json:"matched_definition_id,omitempty"`
	// MatchedDefinitionLabel is the display label of that definition.
	MatchedDefinitionLabel string `json:"matched_definition_label,omitempty"`
}

// ImportRecordsRequest is the request body for POST /contacts/import/records
// (T96): a batch of neutral Card/CRM records — the same shape the REST
// create/update endpoints accept, which is what Android's device-contacts
// import produces — to be run through the standard import preview pipeline
// (duplicate detection, per-row merge diff, within-batch detection) and then
// confirmed via the shared confirm endpoint.
type ImportRecordsRequest struct {
	Records []ContactRecordInput `json:"records" validate:"required,min=1,max=500"`
}

// ImportPreviewResponse contains the full preview data
type ImportPreviewResponse struct {
	SessionID      string             `json:"session_id"`
	Rows           []ImportRowPreview `json:"rows"`
	TotalRows      int                `json:"total_rows"`
	ValidRows      int                `json:"valid_rows"`
	DuplicateCount int                `json:"duplicate_count"`
	ErrorCount     int                `json:"error_count"`
}

// RowImportAction specifies what to do with each row
type RowImportAction struct {
	RowIndex int    `json:"row_index" validate:"min=0"`
	Action   string `json:"action" validate:"required,oneof=skip add update"`
}

// ImportConfirmRequest is sent to execute the import
type ImportConfirmRequest struct {
	SessionID string            `json:"session_id" validate:"required"`
	Actions   []RowImportAction `json:"actions" validate:"required,dive"`

	// FieldMappings carries the issue #514 custom-field promotion decisions for
	// this confirm — one entry per discovered X-* property, telling the
	// confirm path to promote it into an existing or newly-created
	// FieldDefinition ("map"/"create") or leave it in passthrough ("ignore").
	// Absent/empty means every discovered X-* property keeps its passthrough
	// default except those whose name matches an existing definition's vcard
	// projection, which auto-promote (the export round-trip guarantee). Only
	// VCF/JSContact/records (VCF-like) sessions consume these — CSV sessions
	// have no discovered properties and send nothing.
	FieldMappings []ImportFieldMapping `json:"field_mappings,omitempty"`
}

// ImportFieldMapping is one decision from the issue #514 wizard step: what to
// do with a discovered X-* vCard/JSContact property during confirm. It
// mirrors the CSV column-mapping wizard's ColumnMapping in spirit — a
// per-property decision that either routes data into a first-class field or
// leaves it untouched — except the target is a FieldDefinition rather than a
// flat contact field, and the default ("ignore") preserves passthrough
// fidelity rather than discarding data.
type ImportFieldMapping struct {
	// PropertyName is the discovered X-* property this decision applies to
	// (matches DiscoveredProperty.Name case-insensitively).
	PropertyName string `json:"property_name" validate:"required"`
	// Action is "ignore" (keep in passthrough, the default), "map" (promote
	// into the FieldDefinition named by FieldDefinitionID), or "create"
	// (promote into a newly-created definition with Label/Type; the key and
	// vcard projection are derived from the property name).
	Action string `json:"action" validate:"required,oneof=ignore map create"`
	// FieldDefinitionID is required when Action == "map": the existing
	// definition (owned by the current user) to promote values into.
	FieldDefinitionID string `json:"field_definition_id"`
	// Label is used when Action == "create": the new definition's display
	// label. Defaults to a title-cased derivation of the property name.
	Label string `json:"label"`
	// Type is used when Action == "create": the new definition's field type.
	// Defaults to "text" (accepts any vCard value).
	Type string `json:"type" validate:"omitempty,oneof=string text number boolean date datetime uri email phone enum"`
}

// ImportResult summarizes what happened during import
type ImportResult struct {
	TotalProcessed int      `json:"total_processed"`
	Created        int      `json:"created"`
	Updated        int      `json:"updated"`
	Skipped        int      `json:"skipped"`
	Errors         []string `json:"errors"`
}

// ImportSession stores temporary import data server-side (not persisted to DB)
type ImportSession struct {
	ID        string
	UserID    uint
	Headers   []string
	Rows      [][]string
	CreatedAt time.Time
	ExpiresAt time.Time
	// Cached preview data (set after PreviewImport is called)
	Mappings      []ColumnMapping
	PreviewRows   []ImportRowPreview
	PreviewCached bool
}

File: frontend/src/api/import.ts
import { API_BASE_URL, apiFetch, getAuthHeaders, parseErrorResponse } from './client';

// Column mapping between CSV column and contact field
export interface ColumnMapping {
  csv_column: string;
  contact_field: string;
  // Multi-value entry index (0-based). Ties a value column to its label/parts within the
  // same logical entry (e.g. "E-mail 1 - Value" + "E-mail 1 - Label")
  group: number;
}

// Response from CSV upload
export interface ImportUploadResponse {
  session_id: string;
  headers: string[];
  suggested_mappings: ColumnMapping[];
  row_count: number;
  sample_data: string[][];
}

// Match info for duplicate detection
export interface DuplicateMatch {
  existing_contact_id: number;
  existing_firstname: string;
  existing_lastname: string;
  existing_email: string;
  existing_phone: string;
  match_reason: 'name' | 'email' | 'phone';
}

// One scalar field the "Merge" (update) action will overwrite on the matched
// existing contact (T96). Field is a stable camelCase key; label is the
// human-readable fallback.
export interface ImportScalarChange {
  field: string;
  label: string;
  old: string;
  new: string;
}

// One multi-valued entry the "Merge" action will append to the matched
// existing contact. Kind is one of email/phone/address/url/impp.
export interface ImportAddedValue {
  kind: string;
  value: string;
}

// Exactly what "Merge" will change on the matched existing contact: scalars
// overwritten (incoming-wins-when-non-empty) and multi-valued entries
// appended (additive).
export interface ImportMergeDiff {
  updated: ImportScalarChange[];
  added: ImportAddedValue[];
}

// One degradation-policy event from a format adapter, in the same structured
// severity/concept/message shape the export loss reports use (DATA-02, issue
// #442). Mirrors backend contactmodel.Diagnostic.
export interface ImportDiagnostic {
  severity: 'warn' | 'info';
  concept?: string;
  message: string;
}

// Preview row with parsed contact and status
export interface ImportRowPreview {
  row_index: number;
  parsed_contact: Record<string, string>;
  validation_errors: string[];
  duplicate_match: DuplicateMatch | null;
  suggested_action: 'add' | 'skip' | 'update';
  // T96: present exactly when duplicate_match is -- what "Merge" would change.
  merge_diff: ImportMergeDiff | null;
  // T96: when present, the row_index of an earlier row in the SAME import that
  // this row duplicates. Such rows default to "skip".
  batch_duplicate_of: number | null;
  // Issue #514: the unknown X-* vCard/JSContact properties discovered on this
  // row that can be promoted to custom fields. Absent when the row carried none.
  custom_field_candidates?: DiscoveredCustomProperty[];
  // DATA-02 (issue #442): adapter degradation diagnostics for this row
  // (empty/absent for CSV/records-import rows, which don't go through an
  // adapter).
  diagnostics?: ImportDiagnostic[];
}

// One unknown X-* property found on an imported contact that can be promoted
// to a custom field (issue #514). Mirrors backend models.DiscoveredProperty.
export interface DiscoveredCustomProperty {
  // The vCard property name as stored in passthrough (lowercase, e.g.
  // "x-favorite-color").
  name: string;
  // Human-readable rendering of the property's value.
  value: string;
  // When present, an existing FieldDefinition whose vcard:X-<NAME> projection
  // matches this property. The wizard pre-selects "map to this definition"
  // and the confirm path auto-promotes absent an override.
  matched_definition_id?: string;
  matched_definition_label?: string;
}

// Response from preview request
export interface ImportPreviewResponse {
  session_id: string;
  rows: ImportRowPreview[];
  total_rows: number;
  valid_rows: number;
  duplicate_count: number;
  error_count: number;
}

// Action for a specific row
export interface RowImportAction {
  row_index: number;
  action: 'skip' | 'add' | 'update';
}

// One decision from the issue #514 "Custom fields" wizard step: promote a
// discovered X-* property into an existing ("map") or newly-created ("create")
// custom-field definition, or keep it in passthrough ("ignore", the default).
// Mirrors backend models.ImportFieldMapping.
export interface ImportFieldMapping {
  property_name: string;
  action: 'ignore' | 'map' | 'create';
  // Required when action === 'map'.
  field_definition_id?: string;
  // Used when action === 'create': the new definition's label. Defaults to a
  // title-cased derivation of the property name server-side.
  label?: string;
  // Used when action === 'create': the new definition's type. Defaults to
  // 'text' server-side.
  type?: string;
}

// Final import result
export interface ImportResult {
  total_processed: number;
  created: number;
  updated: number;
  skipped: number;
  errors: string[];
}

// One persisted import outcome (issue #651). Mirrors backend models.ImportRun
// and migration 000042's `format` CHECK vocabulary — if a token is added
// backend-side, add it here too (no dynamic type-list endpoint by design).
export interface ImportRun {
  id: number;
  format: 'csv' | 'vcf' | 'jscontact' | 'records';
  total_processed: number;
  created: number;
  updated: number;
  skipped: number;
  error_count: number;
  created_at: string;
}

// Contact fields that can be imported (mirrors backend models.ImportableContactFields)
export const IMPORTABLE_CONTACT_FIELDS = [
  // Name
  'firstname',
  'lastname',
  'middle_name',
  'prefix',
  'suffix',
  'nickname',
  'gender',
  // Dates
  'birthday',
  'anniversary',
  // Multi-value: email / phone
  'email',
  'email_label',
  'phone',
  'phone_label',
  // Multi-value: address parts
  'address_street',
  'address_city',
  'address_region',
  'address_postal',
  'address_country',
  'address_label',
  // Multi-value: web / IM
  'url',
  'url_label',
  'impp',
  'impp_label',
  // Organization
  'organization',
  'department',
  'job_title',
  'role',
  // Free-text
  'how_we_met',
  'work_information',
  'contact_information',
  // Groupings. Distinct targets since T3: 'circles' creates Circle +
  // membership rows, 'tags' creates Tag + assignment rows. Both were missing
  // from this list entirely, so the mapping dropdown could not offer them
  // even though the backend accepted 'circles'.
  'circles',
  'tags',
] as const;

export const REPEATABLE_VALUE_FIELDS = new Set<string>(['email', 'phone', 'url', 'impp']);

// Human-readable labels for contact fields
export const CONTACT_FIELD_LABELS: Record<string, string> = {
  firstname: 'First Name',
  lastname: 'Last Name',
  middle_name: 'Middle Name',
  prefix: 'Name Prefix',
  suffix: 'Name Suffix',
  nickname: 'Nickname',
  gender: 'Gender',
  birthday: 'Birthday',
  anniversary: 'Anniversary',
  email: 'Email',
  email_label: 'Email – Type',
  phone: 'Phone',
  phone_label: 'Phone – Type',
  address_street: 'Address – Street',
  address_city: 'Address – City',
  address_region: 'Address – Region/State',
  address_postal: 'Address – Postal Code',
  address_country: 'Address – Country',
  address_label: 'Address – Type',
  url: 'Website',
  url_label: 'Website – Type',
  impp: 'IM / Social',
  impp_label: 'IM / Social – Type',
  organization: 'Organization',
  department: 'Department',
  job_title: 'Job Title',
  role: 'Role',
  how_we_met: 'How We Met',
  work_information: 'Work Information',
  contact_information: 'Contact Information',
  circles: 'Circles',
  tags: 'Tags',
};

// Upload a CSV file for import
export async function uploadCSVForImport(file: File): Promise<ImportUploadResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await apiFetch(`${API_BASE_URL}/contacts/import/upload`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Get import preview with applied mappings
export async function getImportPreview(
  sessionId: string,
  mappings: ColumnMapping[],
): Promise<ImportPreviewResponse> {
  const response = await apiFetch(`${API_BASE_URL}/contacts/import/preview`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify({
      session_id: sessionId,
      mappings,
    }),
  });

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Confirm and execute the import (CSV)
export async function confirmImport(
  sessionId: string,
  actions: RowImportAction[],
): Promise<ImportResult> {
  const response = await apiFetch(`${API_BASE_URL}/contacts/import/confirm`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify({
      session_id: sessionId,
      actions,
    }),
  });

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Upload a VCF file for import (returns preview directly, no mapping needed)
export async function uploadVCFForImport(file: File): Promise<ImportPreviewResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await apiFetch(`${API_BASE_URL}/contacts/import/vcf/upload`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Recent import outcomes for the current user, newest first (issue #651).
export async function getImportHistory(): Promise<ImportRun[]> {
  const response = await apiFetch(`${API_BASE_URL}/contacts/import/history`, {
    method: 'GET',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Confirm and execute VCF import (with photo processing)
export async function confirmVCFImport(
  sessionId: string,
  actions: RowImportAction[],
  fieldMappings: ImportFieldMapping[] = [],
): Promise<ImportResult> {
  const response = await apiFetch(`${API_BASE_URL}/contacts/import/vcf/confirm`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify({
      session_id: sessionId,
      actions,
      field_mappings: fieldMappings,
    }),
  });

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}
